src/execution/sema/sema_builtin.cpp

```diff
@@ -1840,6 +1840,83 @@ void Sema::CheckMathTrigCall(ast::CallExpr *call, ast::Builtin builtin) {
   call->SetType(GetBuiltinType(return_kind));
 }
 
+void Sema::CheckAtomicCall(ast::CallExpr *call, ast::Builtin builtin) {
+  const auto &call_args = call->Arguments();
+
+  int arg_count;
+  ast::BuiltinType::Kind operand_kind;
+  switch (builtin) {
+    case ast::Builtin::AtomicAnd1:
+    case ast::Builtin::AtomicOr1:
+      arg_count = 2;
+      operand_kind = ast::BuiltinType::Uint8;
+      break;
+    case ast::Builtin::AtomicAnd2:
+    case ast::Builtin::AtomicOr2:
+      arg_count = 2;
+      operand_kind = ast::BuiltinType::Uint16;
+      break;
+    case ast::Builtin::AtomicAnd4:
+    case ast::Builtin::AtomicOr4:
+      arg_count = 2;
+      operand_kind = ast::BuiltinType::Uint32;
+      break;
+    case ast::Builtin::AtomicAnd8:
+    case ast::Builtin::AtomicOr8:
+      arg_count = 2;
+      operand_kind = ast::BuiltinType::Uint64;
+      break;
+    case ast::Builtin::AtomicCompareExchange1:
+      arg_count = 3;
+      operand_kind = ast::BuiltinType::Uint8;
+      break;
+    case ast::Builtin::AtomicCompareExchange2:
+      arg_count = 3;
+      operand_kind = ast::BuiltinType::Uint16;
+      break;
+    case ast::Builtin::AtomicCompareExchange4:
+      arg_count = 3;
+      operand_kind = ast::BuiltinType::Uint32;
+      break;
+    case ast::Builtin::AtomicCompareExchange8:
+      arg_count = 3;
+      operand_kind = ast::BuiltinType::Uint64;
+      break;
+    default:
+      UNREACHABLE("Impossible atomic call");
+  }
+
+  if (!CheckArgCount(call, arg_count)) return;
+
+  if (arg_count == 2) {
+    if (!IsPointerToSpecificBuiltin(call_args[0]->GetType(), operand_kind)) {
+      ReportIncorrectCallArg(call, 0, GetBuiltinType(operand_kind));
+      return;
+    }
+    if (!call_args[1]->GetType()->IsSpecificBuiltin(operand_kind)) {
+      ReportIncorrectCallArg(call, 1, GetBuiltinType(operand_kind));
+      return;
+    }
+  } else {
+    // TODO(John): The typing for this should probably be relaxed some to
+    // allow CAS on any type that is of a supported width.
+    if (!IsPointerToSpecificBuiltin(call_args[0]->GetType(), operand_kind)) {
+      ReportIncorrectCallArg(call, 0, GetBuiltinType(operand_kind));
+      return;
+    }
+    if (!IsPointerToSpecificBuiltin(call_args[1]->GetType(), operand_kind)) {
+      ReportIncorrectCallArg(call, 0, GetBuiltinType(operand_kind));
+      return;
+    }
+    if (!call_args[2]->GetType()->IsSpecificBuiltin(operand_kind)) {
+      ReportIncorrectCallArg(call, 1, GetBuiltinType(operand_kind));
+      return;
+    }
+  }
+
+  call->SetType(GetBuiltinType(ast::BuiltinType::Nil));
```

Comment:
I thought CAS usually returned a bool, but if we don't need it then I guess it is fine / can be added later.